### PR TITLE
Add support for properties, JSON and YAML configuration in Core SDK

### LINKED ISSUE #597: Enhance Core SDK to support .properties configuration
The Core SDK should allow users to specify `.properties` in the following locations to override defaults.

1. `classpath:com/stormpath/sdk/config/stormpath.properties`
2. `classpath:stormpath.properties`
3. `$HOME/.stormpath/apiKey.properties`
4. `$HOME/.stormpath/stormpath.properties`
5. `System.getProperty("user.dir") + File.separator + "apiKey.properties"`
6. `System.getProperty("user.dir") + File.separator + "stormpath.properties`

NOTE: The first file does not currently exist, so it will need to be created with all the [defaults](https://github.com/stormpath/stormpath-sdk-spec/blob/master/specifications/config.md#default-configuration).

See https://github.com/stormpath/stormpath-sdk-java/wiki/Overriding-Default-Configuration for additional information.

### PR DESCRIPTION
**Pull request by mraible**
This fixes #597 and #588 and #598. 

There's still more to do on 588 related to all the properties I added to `extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json` from `web.stormpath.properties`. Specifically, all the descriptions and default values still need to be populated.

I'm creating this PR now in hopes of getting it reviewed before I continue the tedious work of populating descriptions and default values. 😉

### CODE REVIEW

File: extensions/servlet/src/test/groovy/com/stormpath/sdk/servlet/config/SpecConfigVersusWebPropertiesTest.groovy [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+import static org.testng.Assert.assertEquals
+
+/**
+ * A test that downloads the web config from the spec and compares
+ * it with com/stormpath/sdk/servlet/config/web.stormpath.properties
+ *
+ * @since 1.0.0
+ */
+class SpecConfigVersusWebPropertiesTest {
```
Comment (mrioan) [inline extensions/servlet/src/test/groovy/com/stormpath/sdk/servlet/config/SpecConfigVersusWebPropertiesTest.groovy]: nice!

File: extensions/servlet/src/test/groovy/com/stormpath/sdk/servlet/config/SpecConfigVersusWebPropertiesTest.groovy [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        }
+
+        if (diff.size > 0) {
+            println "Missing keys in default config:"
+            diff.each {
+                println "${it}"
+            }
+        }
+
+        assertEquals diff.size(), 0, "Missing keys in default config: ${diff}"
```
Comment (mrioan) [inline extensions/servlet/src/test/groovy/com/stormpath/sdk/servlet/config/SpecConfigVersusWebPropertiesTest.groovy]: Could you please add a clear message of why this error arose? Builds will start failing out of the blue as soon as someone changes the spec. Please add something like "It is very likely that a new property was added in XX and the Java SDK is missing it. It must be supported in the Java SDK in order for the build to pass.  In the meantime you could allow the build to pass by adding the missing property to the NOT_YET_IMPLEMENTED list but be sure to remember to addd support for it in the near term".

Comment (mrioan) [inline extensions/servlet/src/test/groovy/com/stormpath/sdk/servlet/config/SpecConfigVersusWebPropertiesTest.groovy]: Could you please add a clear message of why this error arose? Builds will start failing out of the blue as soon as someone changes the spec. Please add something like "It is very likely that a new property was added in XX and the Java SDK is missing it. It must be supported in the Java SDK in order for the build to pass.  In the meantime you could allow the build to pass by adding the missing property to the NOT_YET_IMPLEMENTED list but be sure to remember to addd support for it in the near term".

Comment (mrioan) [inline extensions/servlet/src/test/groovy/com/stormpath/sdk/servlet/config/SpecConfigVersusWebPropertiesTest.groovy]: Could you please add a clear message of why this error arose? Builds will start failing out of the blue as soon as someone changes the spec. Please add something like "It is very likely that a new property was added in XX and the Java SDK is missing it. It must be supported in the Java SDK in order for the build to pass.  In the meantime you could allow the build to pass by adding the missing property to the NOT_YET_IMPLEMENTED list but be sure to remember to addd support for it in the near term".

File: extensions/servlet/src/test/groovy/com/stormpath/sdk/servlet/config/SpecConfigVersusWebPropertiesTest.groovy [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    @Test
+    void verifyPropertiesInDefaultAreInSpec() {
+        def diff = defaultProperties.findResults { k,v ->
+            specProperties.containsKey(k) ? null : k
+        }
+
+        assertEquals diff.size(), 66, "Missing keys in spec config: ${diff}"
```
Comment (mrioan) [inline extensions/servlet/src/test/groovy/com/stormpath/sdk/servlet/config/SpecConfigVersusWebPropertiesTest.groovy]: Could you please add a clear message of why this error arose? Builds will start failing out of the blue as soon as someone changes the spec. Please add something like "It is very likely that a new property was added in XX and the Java SDK is missing it. It must be supported in the Java SDK in order for the build to pass.  In the meantime you could allow the build to pass by adding the missing property to the NOT_YET_IMPLEMENTED list but be sure to remember to addd support for it in the near term".

File: extensions/spring/boot/stormpath-spring-boot-starter/src/test/groovy/com/stormpath/spring/boot/config/SpringMetadataVersusCorePropertiesTest.groovy [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import com.stormpath.sdk.impl.config.ResourcePropertiesSource
+import com.stormpath.sdk.impl.io.ClasspathResource
+import com.stormpath.sdk.impl.io.DefaultResourceFactory
+import com.stormpath.sdk.impl.io.Resource
+import com.stormpath.sdk.impl.io.ResourceFactory
+import org.testng.annotations.BeforeClass
+import org.testng.annotations.Test
+
+import static org.testng.Assert.assertEquals
+/**
```
Comment (mrioan) [inline extensions/spring/boot/stormpath-spring-boot-starter/src/test/groovy/com/stormpath/spring/boot/config/SpringMetadataVersusCorePropertiesTest.groovy]: Blank line please

File: extensions/spring/boot/stormpath-spring-boot-starter/src/test/groovy/com/stormpath/spring/boot/config/SpringMetadataVersusCorePropertiesTest.groovy [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ * A test that compares the properties in com/stormpath/sdk/config/stormpath.properties
+ * with the ones defined in additional-spring-configuration-metadata.json.
+ *
+ * @since 1.0.0
+ */
+class SpringMetadataVersusCorePropertiesTest {
+    def METADATA_CONFIG_LOCATION = ClasspathResource.SCHEME_PREFIX + "META-INF/additional-spring-configuration-metadata.json"
```
Comment (mrioan) [inline extensions/spring/boot/stormpath-spring-boot-starter/src/test/groovy/com/stormpath/spring/boot/config/SpringMetadataVersusCorePropertiesTest.groovy]: idem here

File: extensions/spring/boot/stormpath-spring-boot-starter/src/test/groovy/com/stormpath/spring/boot/config/SpringMetadataVersusCorePropertiesTest.groovy [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import com.stormpath.sdk.impl.io.Resource
+import com.stormpath.sdk.impl.io.ResourceFactory
+import org.testng.annotations.BeforeClass
+import org.testng.annotations.Test
+
+import static org.testng.Assert.assertEquals
+/**
+ * A test that compares the properties in com/stormpath/sdk/config/stormpath.properties
+ * with the ones defined in additional-spring-configuration-metadata.json.
+ *
```
Comment (mrioan) [inline extensions/spring/boot/stormpath-spring-boot-starter/src/test/groovy/com/stormpath/spring/boot/config/SpringMetadataVersusCorePropertiesTest.groovy]: Could you please addd a comment saying something like "This test will alert when a new property has been added to the framework but it was not documented in `additional-spring-configuration-metadata.json`" ?

File: extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json [hunk 7]
```diff
@@ -1,6 +1,9 @@
 {
   "groups": [
     {
+      "name": "stormpath.application"
+    },
+    {
```
Comment (mrioan) [inline extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json]: I think these properties should be documented, should't them?

File: extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json [hunk 8]
```diff
@@ -482,22 +473,899 @@
...
     },
     {
       "name": "stormpath.web.application.domain",
       "type": "java.lang.String",
       "description": "The base domain of your application. For example if your app resides in 'myapp.mycompany.com', your base domain will be 'mycompany.com'.",
       "defaultValue": "Inferred based on heuristics by default. However if your application is not deployed to an apex domain, like myapp.com, you *must* specify your application's base domain, e.g. 'myapp.mycompany.com'."
+    },
+    {
+      "name": "stormpath.application.href",
+      "type": "java.lang.String",
```
Comment (mrioan) [inline extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json]: Missing description and default value in some of these...

File: impl/src/main/java/com/stormpath/sdk/impl/config/ClientConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * This class holds the default configuration properties
+ *
+ * @since 1.0
+ */
+public class ClientConfiguration {
+    private String apiKeyFile;
```
Comment (mrioan) [inline impl/src/main/java/com/stormpath/sdk/impl/config/ClientConfiguration.java]: blank line please

File: impl/src/main/java/com/stormpath/sdk/impl/config/ClientConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+import com.stormpath.sdk.cache.CacheConfigurationBuilder;
+import com.stormpath.sdk.client.AuthenticationScheme;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * This class holds the default configuration properties
+ *
```
Comment (mrioan) [inline impl/src/main/java/com/stormpath/sdk/impl/config/ClientConfiguration.java]: Is this accurate? I would think that this class is a placeholder for all the configuration properties, both default ones and overwritten by the user. If so I would add a comment saying something like "During application initialization all the properties found in the pre-defined locations that are defined by the user will be added here in the order defined in XXX. Unset value will keep default values pre-defined in YYY"

XXX: is a link to either the spec, the github issue  or the wiki page that clearly is documenting this order
YYY: link to the file in the Java SDK containing the default values

File: extensions/spring/boot/stormpath-spring-boot-starter/src/test/groovy/com/stormpath/spring/boot/config/SpringMetadataVersusCorePropertiesTest.groovy (hunk 4 above)
Comment (mraible) [inline extensions/spring/boot/stormpath-spring-boot-starter/src/test/groovy/com/stormpath/spring/boot/config/SpringMetadataVersusCorePropertiesTest.groovy]: Just so I know what you're asking - you're saying there should be a blank line before the javadoc, correct?

File: extensions/spring/boot/stormpath-spring-boot-starter/src/test/groovy/com/stormpath/spring/boot/config/SpringMetadataVersusCorePropertiesTest.groovy (hunk 5 above)
Comment (mraible) [inline extensions/spring/boot/stormpath-spring-boot-starter/src/test/groovy/com/stormpath/spring/boot/config/SpringMetadataVersusCorePropertiesTest.groovy]: Blank line after the class?

File: extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json (hunk 7 above)
Comment (mraible) [inline extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json]: I'm not sure if the groups have a `description` field. @lhazlewood I assumed these groups were only necessary for keys that had a number of properties under the same "group". Is that correct?

File: extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json (hunk 8 above)
Comment (mraible) [inline extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json]: You are correct. I copied the descriptions and defaultValues from `web.stormpath.properties` (if they existed). @dogeared added issue #620 for the rest.

**Comment by mrioan** [comment]
I would like to see a few tests that demonstrate that the overrides are happening in the right order.

Back to you @mraible

**Comment by mraible** [comment]
@mrioan I fixed Javadoc issues, improved test failure messages, fixed the syntax issues you mentioned and improved [one test](https://github.com/stormpath/stormpath-sdk-java/pull/617/commits/c8fd0d7e0ab15324bb031b988e5683a5ef67c271) to prove overrides are working. If you'd like additional tests, please let me know. 

I could create `~/stormpath.*` files on-the-fly for more rigorous tests. Unfortunately, I've found it difficult to delete these files while the tests are running - so other tests will pick up this configuration. That's why I didn't do it this way.

File: extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json (hunk 8 above)
Comment (mrioan) [inline extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/resources/META-INF/additional-spring-configuration-metadata.json]: Great

**Comment by mrioan** [comment]
@mraible LGTM. If you think @lhazlewood or @dogeared should take a look at it as well for some reason then contact them. Otherwise you are good to merge. Back to you. Thanks

**Pull request merged**